### PR TITLE
Simplify XContent detection.

### PR DESCRIPTION
**Pull request by jpountz**
Currently we duplicate the logic for BytesReferences and InputStreams.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java [hunk 1]
```diff
@@ -216,65 +215,16 @@ public static XContentType xContentType(InputStream si) throws IOException {
...
-                if (val == -1) {
-                    return null;
-                }
-                if (val == '{') {
-                    return XContentType.JSON;
-                }
-                if (Character.isWhitespace(val) == false) {
+            final byte[] firstBytes = new byte[GUESS_HEADER_LENGTH];
+            int read = 0;
+            while (read < firstBytes.length) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java]: can we use org.elasticsearch.common.io.Streams#readFully(java.io.InputStream, byte[]) ?

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java]: good point!

**Comment by bleskes** [comment]
+1 on share the code. I wonder if we should flip this around and use the stream input as the basic implementation and use org.elasticsearch.common.bytes.BytesReference#streamInput  to call it from the bytes reference variant. It's just slighty more efficient as it doesn't read ahead.

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java]: I wonder if we should this api to the forbidden api pointing people at the utilities...  (like we did in the write variants).

**Comment by jpountz** [comment]
I wondered about the same but using the stream input version as a basic implementation had the downside of being a bit harder to read (because it had to check every read against -1 while the BytesReference version can just check the length once) as well as require the stream input returned by every BytesReference to support mark/reset.

**Comment by s1monw** [comment]
lets not over-design this - simplicity is king here

**Comment by bleskes** [comment]
yeah, I hear you. I'm not sure if it's that's easier to read but we're splitting hairs :)  one up side of keeping the stream input variant as the basis is that it might be easier to convert org.elasticsearch.common.xcontent.XContentFactory#xContentType(java.lang.CharSequence)  as well (wrapping the char sequence  with a InputStreams)

**Comment by bleskes** [comment]
so bottom line - up to you :)

**Pull request merged**